src/Cli/dotnet/commands/VirtualProjectBuildingCommand.cs

```diff
@@ -0,0 +1,194 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#nullable enable
+
+using System.Collections.Immutable;
+using System.Xml;
+using Microsoft.Build.Construction;
+using Microsoft.Build.Definition;
+using Microsoft.Build.Evaluation;
+using Microsoft.Build.Execution;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Logging;
+using Microsoft.DotNet.Cli.Utils;
+using Microsoft.DotNet.Tools.Run;
+
+namespace Microsoft.DotNet.Tools;
+
+/// <summary>
+/// Used to build a virtual project file in memory to support <c>dotnet run file.cs</c>.
+/// </summary>
+internal sealed class VirtualProjectBuildingCommand
+{
+    public Dictionary<string, string> GlobalProperties { get; } = new(StringComparer.OrdinalIgnoreCase);
+    public required string EntryPointFileFullPath { get; init; }
+
+    public int Execute(string[] binaryLoggerArgs, LoggerVerbosity verbosity)
+    {
+        var binaryLoggers = GetBinaryLoggers(binaryLoggerArgs);
+        var consoleLogger = new ConsoleLogger(verbosity);
+        Dictionary<string, string?> savedEnvironmentVariables = new();
+        try
+        {
+            // Set environment variables.
+            foreach (var (key, value) in MSBuildForwardingAppWithoutLogging.GetMSBuildRequiredEnvironmentVariables())
+            {
+                savedEnvironmentVariables[key] = Environment.GetEnvironmentVariable(key);
+                Environment.SetEnvironmentVariable(key, value);
+            }
+
+            // Setup MSBuild.
+            var projectCollection = new ProjectCollection(
+                GlobalProperties,
+                [.. binaryLoggers, consoleLogger],
+                ToolsetDefinitionLocations.Default);
+            var parameters = new BuildParameters(projectCollection)
+            {
+                Loggers = projectCollection.Loggers,
+                LogTaskInputs = binaryLoggers.Length != 0,
+            };
+            BuildManager.DefaultBuildManager.BeginBuild(parameters);
+
+            // Do a restore first (equivalent to MSBuild's "implicit restore", i.e., `/restore`).
+            // See https://github.com/dotnet/msbuild/blob/a1c2e7402ef0abe36bf493e395b04dd2cb1b3540/src/MSBuild/XMake.cs#L1838
+            // and https://github.com/dotnet/msbuild/issues/11519.
+            var restoreRequest = new BuildRequestData(
```

Comment:
Instead of doing any of this, why can't we make a fake project then "call" RunCommand.Execute (with modified parameters) on it? That would make me feel much more comfortable.

Reply:
1. Performance is presumably better by invoking MSBuild in memory. We are planning to do more optimizations as well (see the design doc), e.g., skipping MSBuild invocation and invoking just csc.exe or even just the target dll. For that, we will need to communicate with MSBuild to determine the csc arguments.
2. If we create a fake project, it will show up on disk during the build time, which is not a great user experience. We could create it inside `obj` folder but that's a problem because then it has a different path which is an observable difference.

Reply:
When I wrote that comment, I'd glanced only very briefly at RunCommand, and many of its methods operated on ProjectInstances, so my question was whether you could create a ProjectInstance and pass it to RunCommand.

I looked further just now, and though it does a lot of checks with ProjectInstances, it ultimately creates an MSBuildForwardingApp that calls MSBuild's Main method directly, and there's no overload for that that takes in a ProjectInstance.

I will caution against worrying too much about performance at this stage. RunCommand does a lot of work that you don't, and I'd rather this be correct than fast. If this is correct, why is RunCommand bulky? Could RunCommand be cut down and perhaps even made more performant if it were to follow a model similar to yours?

Reply:
As far as I understand, RunCommand only uses ProjectInstances and MSBuild APIs to get the RunArguments (users can use that to override how their project actually runs, e.g., what command line arguments are passed to it). This PR reuses all that. For everything else, RunCommand just calls MSBuild's Main method. We cannot reuse that, there is no way to pass XML project file text to MSBuild via its Main method as far as I know.

Reply:
> RunCommand does a lot of work that you don't

Can you elaborate on that? I'm not sure what work RunCommand does that this PR doesn't.

Note that this PR just hooks into the "build" stage of RunCommand. The rest that does execution is still the same.

Reply:
> * Launch settings (which we aren't supporting here intentionally)

We support those now (although we previously didn't, you are right). The support for that is shared because it's part of the "run" phase (after the "build" phase) which is the same between both project-based and file-based projects.